File: lib/src/main/java/com/example/lib/geeksforgeeks/ds/binaryTree/VerticalOrderTraversal.java
package com.example.lib.geeksforgeeks.ds.binaryTree;

import java.util.Iterator;
import java.util.LinkedList;
import java.util.Map;
import java.util.TreeMap;

/**
 * Created by vihaanverma on 03/12/17.
 */

public class VerticalOrderTraversal {
    public static void main(String[] args) {


        Node node = new Node(1);

        node.addLeftNode(new Node(2));
        node.addRightNode(new Node(3));

        node.getLeftNode().addLeftNode(new Node(4));
        node.getLeftNode().addRightNode(new Node(5));

        node.getRightNode().addLeftNode(new Node(6));
        node.getRightNode().addRightNode(new Node(7));


        node.getRightNode().getLeftNode().addRightNode(new Node(8));

        node.getRightNode().getRightNode().addRightNode(new Node(9));

        VerticalOrderTraversal verticalOrderTraversal = new VerticalOrderTraversal();
        verticalOrderTraversal.verticalTraversal(node,0);
        verticalOrderTraversal.printList();
    }

    Map<Integer, LinkedList<Integer>> map = new TreeMap<>();
    private void verticalTraversal(Node node, int depth)
    {
        if(node==null)
        {
            return ;
        }

        LinkedList<Integer> list;
        if(map.containsKey(depth))
        {
             list = map.get(depth);
        }
        else{
            list = new LinkedList<>();
            map.put(depth, list);
        }

        list.add(node.getValue());

        verticalTraversal(node.getLeftNode(), depth-1);
        verticalTraversal(node.getRightNode(), depth+1);
    }

    private void printList(){

        for(Map.Entry<Integer, LinkedList<Integer>> entry : map.entrySet())
        {
            int key = entry.getKey();
            LinkedList list = entry.getValue();
            Iterator it = list.iterator();
            System.out.print(key + " - ");
            while(it.hasNext())
            {
                System.out.print(it.next()+ ",");
            }
            System.out.println();
        }
    }

}

File: lib/src/main/java/com/example/lib/geeksforgeeks/binaryTree/VerticalOrderTraversal.kt
package com.example.lib.geeksforgeeks.binaryTree

import java.util.*


fun main() {

    val root = Node(1)
    val binaryTree = BinaryTree(root)
    root.leftNode = Node(2)
    root.rightNode = Node(3)
    root.leftNode.leftNode = Node(4)
    root.leftNode.rightNode = Node(5)
    root.rightNode.leftNode = Node(6)
    root.rightNode.rightNode = Node(7)
    root.rightNode.leftNode.rightNode = Node(8)
    root.rightNode.rightNode.rightNode = Node(9)
    verticalTraversal(root, 0)
    println("Vertical Order traversal is")
    printVerticalOrder(root)
}


val map = TreeMap<Int, ArrayList<Node>>()
fun verticalTraversal(node: Node, verticalCount: Int) {

    if (map.containsKey(verticalCount)) {
        val arrayList = map[verticalCount]
        arrayList?.add(node)
    } else {
        val arrayListOf = arrayListOf<Node>()
        map[verticalCount] = arrayListOf
        arrayListOf.add(node)
    }
    if (node.getLeftNode() != null) {
        verticalTraversal(node.leftNode, verticalCount - 1)
    }

    if (node.getRightNode() != null) {
        verticalTraversal(node.rightNode, verticalCount + 1)
    }
}

fun printVerticalOrder(root: Node) {
    for (mutableEntry in map) {
        val key = mutableEntry.key
        val value = mutableEntry.value
        val joinToString = value.joinToString(",")
        println("Level $key : $joinToString")
    }
}